storagemarket/provider.go

```diff
@@ -204,40 +204,51 @@ func (p *Provider) GetAsk() *storagemarket.StorageAsk {
 	}
 }
 
+// MakeOfflineDealWithData is called when the Storage Provider imports data for
+// an offline deal (the deal must already have been proposed by the client)
 func (p *Provider) MakeOfflineDealWithData(dealUuid uuid.UUID, filePath string) (pi *api.ProviderDealRejectionInfo, handler *dealHandler, err error) {
-	p.dealLogger.Infow(dealUuid, "execute offline deal called", "filepath", filePath)
+	p.dealLogger.Infow(dealUuid, "import data for offline deal", "filepath", filePath)
 
 	// db should already have a deal with this uuid as the deal proposal should have been agreed before hand
 	ds, err := p.dealsDB.ByID(p.ctx, dealUuid)
 	if err != nil {
-		return nil, nil, fmt.Errorf("no pre-existing deal proposal for offline deal: %w", err)
+		if xerrors.Is(err, sql.ErrNoRows) {
+			return nil, nil, fmt.Errorf("no pre-existing deal proposal for offline deal %s: %w", dealUuid, err)
+		}
+		return nil, nil, fmt.Errorf("getting offline deal %s: %w", dealUuid, err)
 	}
 	if !ds.IsOffline {
-		return nil, nil, errors.New("deal for the given id is not an offline deal")
+		return nil, nil, fmt.Errorf("deal %s is not an offline deal", dealUuid)
+	}
+	if ds.Checkpoint > dealcheckpoints.Accepted {
+		return nil, nil, fmt.Errorf("deal %s has already been imported and reached checkpoint %s", dealUuid, ds.Checkpoint)
 	}
+
 	ds.InboundFilePath = filePath
 
-	// setup clean-up code
-	var dh *dealHandler
+	// create new deal handler
+	dh, isNew := p.mkAndInsertDealHandler(dealUuid)
 	cleanup := func() {
-		if dh != nil {
+		// if the deal handler didn't already exist, clean it up on error
+		if isNew {
```

Comment:
I'm not sure I understand why we need this. Let's discuss.

Reply:
When the deal is proposed we create a deal handler and publish a "new deal" event. Then when the deal is imported we need to get the same deal handler to publish progress events.